File: src/shaders/shader.fs
precision mediump float; // set float to medium precision

// eye location
uniform vec3 uEyePosition; // the eye's position in world
uniform vec3 uLookAt; // vector that indicates where the player is looking

// light properties
uniform vec3 uSunAmbient; // the light's ambient color
uniform vec3 uSunSpecular; // the light's specular color

uniform vec3 uMoonAmbient; // the light's ambient color
uniform vec3 uMoonDiffuse; // the light's diffuse color
uniform vec3 uMoonSpecular; // the light's specular color

uniform vec3 uSunDirection; // the direction of the sun

uniform float uFlashLightOn; // the power/color of the flashLight

// material properties
uniform vec3 uAmbient; // the ambient reflectivity
uniform vec3 uDiffuse; // the diffuse reflectivity
uniform vec3 uSpecular; // the specular reflectivity
uniform float uShininess; // the specular exponent

// // texture properties
// uniform bool uUsingTexture; // if we are using a texture
// uniform sampler2D uTexture; // the texture for the fragment
// varying vec2 vVertexUV; // texture uv of fragment

// geometry properties
varying vec3 vWorldPos; // world xyz of fragment
varying vec3 vVertexNormal; // normal of fragment

void main(void) {

    // ambient term
    vec3 ambient = uAmbient*uSunAmbient;

    // sunset
    vec3 modifiedSun = vec3(0.4 + abs(uSunDirection.y) * 0.6, max(0.5, abs(uSunDirection.y)), max(0.5, min(0.92, abs(uSunDirection.y))));

    // sun diffuse
    vec3 normal = normalize(vVertexNormal);
    vec3 sunLight = uSunDirection;
    float sunLambert = max(0.0,dot(normal,sunLight));
    vec3 sunDiffuse = uDiffuse*modifiedSun*sunLambert; // diffuse from sun

    // moon diffuse
    vec3 moonLight = uSunDirection * -1.0;
    float moonLambert = max(0.0,dot(normal,moonLight));
    vec3 moonDiffuse = uDiffuse*uMoonDiffuse*moonLambert; // diffuse from moon

    // flashlight diffuse
    vec3 eyeOffset = uEyePosition - vWorldPos;
    vec3 eye = normalize(eyeOffset);
    float flashLightLambert = max(0.0,dot(normal,eye));

    // specular term
    vec3 halfVec = normalize(sunLight+eye);
    float highlight = pow(max(0.0,dot(normal,halfVec)),uShininess);
    vec3 specular = uSpecular*uSunSpecular*highlight; // specular term

    // don't use flashLight if angle is too steep
    float angle = dot(uLookAt, eye * -1.0);
    float lightValue = pow(angle, 4.0) * uFlashLightOn / pow(max(1.0, length(eyeOffset)*4.0), 2.0);
    vec3 flashLightDiffuse;
    flashLightDiffuse = uDiffuse*flashLightLambert * lightValue;

    // combine to find lit color
    vec3 litColor = ambient + sunDiffuse + moonDiffuse + flashLightDiffuse;

    gl_FragColor = vec4(litColor, 1.0);
} // end main

File: src/shaders/shader.ts
import { vec3, mat4 } from 'gl-matrix';

export class Shader {
    gl: WebGLRenderingContext;
    vertexShader: WebGLShader;
    fragmentShader: WebGLShader;
    shaderProgram: WebGLProgram;

    public vPosAttribLoc: number;
    public vNormAttribLoc: number;
    public vColorAttribLoc: number;

    public diffuseULoc: WebGLUniformLocation;

    public mMatrixULoc: WebGLUniformLocation;
    public pvmMatrixULoc: WebGLUniformLocation;
    public pvmMatrix: mat4;

    public eyePositionULoc: WebGLUniformLocation;
    public lookAtULoc: WebGLUniformLocation;
    public sunDirectionULoc: WebGLUniformLocation;
    public flashLightOnULoc: WebGLUniformLocation;

    constructor(gl: WebGLRenderingContext,
            fShaderCode: string,
            vShaderCode: string) {
        this.gl = gl;
        this.fragmentShader = gl.createShader(gl.FRAGMENT_SHADER);
        gl.shaderSource(this.fragmentShader, fShaderCode);
        gl.compileShader(this.fragmentShader);

        this.vertexShader = gl.createShader(gl.VERTEX_SHADER);
        gl.shaderSource(this.vertexShader, vShaderCode);
        gl.compileShader(this.vertexShader);

        if (!gl.getShaderParameter(this.fragmentShader, gl.COMPILE_STATUS)) {
            throw "error during fragment shader compile: " + gl.getShaderInfoLog(this.fragmentShader);
        }
        else if (!gl.getShaderParameter(this.vertexShader, gl.COMPILE_STATUS)) {
            throw "error during vertex shader compile: " + gl.getShaderInfoLog(this.vertexShader);
        }

        this.shaderProgram = gl.createProgram();
        gl.attachShader(this.shaderProgram, this.fragmentShader);
        gl.attachShader(this.shaderProgram, this.vertexShader);
        gl.linkProgram(this.shaderProgram);

        if (!gl.getProgramParameter(this.shaderProgram, gl.LINK_STATUS)) { // bad program link
            throw "error during shader program linking: " + gl.getProgramInfoLog(this.shaderProgram);
        }

        gl.useProgram(this.shaderProgram);

        // locate and enable vertex attributes
        this.vPosAttribLoc = gl.getAttribLocation(this.shaderProgram, "aVertexPosition"); // ptr to vertex pos attrib
        gl.enableVertexAttribArray(this.vPosAttribLoc); // connect attrib to array
        this.vNormAttribLoc = gl.getAttribLocation(this.shaderProgram, "aVertexNormal"); // ptr to vertex normal attrib
        gl.enableVertexAttribArray(this.vNormAttribLoc); // connect attrib to array
        this.vColorAttribLoc = gl.getAttribLocation(this.shaderProgram, "aVertexColor"); // ptr to vertex normal attrib
        gl.enableVertexAttribArray(this.vColorAttribLoc); // connect attrib to array

        // locate vertex uniforms
        this.mMatrixULoc = gl.getUniformLocation(this.shaderProgram, "umMatrix"); // ptr to mmat
        this.pvmMatrixULoc = gl.getUniformLocation(this.shaderProgram, "upvmMatrix"); // ptr to pvmmat

        // locate fragment uniforms
        this.eyePositionULoc = gl.getUniformLocation(this.shaderProgram, "uEyePosition"); // ptr to eye position
        this.lookAtULoc = gl.getUniformLocation(this.shaderProgram, "uLookAt"); // ptr to lookAt vector
        this.flashLightOnULoc = gl.getUniformLocation(this.shaderProgram, "uFlashLightOn"); // ptr to toggle flashLight conditional
        let sunAmbientULoc = gl.getUniformLocation(this.shaderProgram, "uSunAmbient"); // ptr to light ambient
        let sunSpecularULoc = gl.getUniformLocation(this.shaderProgram, "uSunSpecular"); // ptr to light specular
        let moonAmbientULoc = gl.getUniformLocation(this.shaderProgram, "uMoonAmbient"); // ptr to light ambient
        let moonDiffuseULoc = gl.getUniformLocation(this.shaderProgram, "uMoonDiffuse"); // ptr to light diffuse
        let moonSpecularULoc = gl.getUniformLocation(this.shaderProgram, "uMoonSpecular"); // ptr to light specular
        this.sunDirectionULoc = gl.getUniformLocation(this.shaderProgram, "uSunDirection"); // ptr to sun direction
        let ambientULoc = gl.getUniformLocation(this.shaderProgram, "uAmbient"); // ptr to ambient
        this.diffuseULoc = gl.getUniformLocation(this.shaderProgram, "uDiffuse"); // ptr to diffuse
        let specularULoc = gl.getUniformLocation(this.shaderProgram, "uSpecular"); // ptr to specular
        let shininessULoc = gl.getUniformLocation(this.shaderProgram, "uShininess"); // ptr to shininess

        // Pass in default values for uniforms
        // var sunDirection = vec3.fromValues(0,1,0);

        var sunAmbient = vec3.fromValues(1, 1, 1); // default light ambient emission
        var sunSpecular = vec3.fromValues(1, 1, 1); // default light specular emission

        var moonAmbient = vec3.fromValues(1, 1, 1); // default light ambient emission
        var moonDiffuse = vec3.fromValues(0.16, 0.18, 0.23); // default light diffuse emission
        var moonSpecular = vec3.fromValues(1, 1, 1); // default light specular emission

        var ambient = vec3.fromValues(0, 0, 0); // default ambient emission
        var diffuse = vec3.fromValues(0.6, 0.6, 0.6); // default diffuse emission
        var specular = vec3.fromValues(0.1, 0.1, 0.1); // default specular emission
        var shininess = 11; // specular exponent

        // var hMatrix = mat4.create(); // handedness matrix
        // var pMatrix = mat4.create(); // projection matrix
        // var vMatrix = mat4.create(); // view matrix
        // var mMatrix = mat4.create(); // model matrix
        // var hpvMatrix = mat4.create(); // hand * proj * view matrices
        // var hpvmMatrix = mat4.create(); // hand * proj * view * model matrices

        // mat4.fromScaling(hMatrix, vec3.fromValues(-1, 1, 1)); // create handedness matrix
        // mat4.perspective(pMatrix, 0.5 * Math.PI, 1, 0.1, 10); // create projection matrix
        // mat4.lookAt(vMatrix, eye, center, up); // create view matrix
        // mat4.multiply(hpvMatrix, hMatrix, pMatrix); // handedness * projection
        // mat4.multiply(hpvMatrix, hpvMatrix, vMatrix); // handedness * projection * view
        // mat4.multiply(hpvmMatrix, hpvMatrix, mMatrix); // handedness * projection * view * model

        // gl.uniformMatrix4fv(this.mMatrixULoc, false, mMatrix); // pass in the m matrix
        // gl.uniformMatrix4fv(this.pvmMatrixULoc, false, hpvmMatrix); // pass in the hpvm matrix

        gl.uniform3fv(sunAmbientULoc, sunAmbient); // pass in the sun's ambient emission
        gl.uniform3fv(sunSpecularULoc, sunSpecular); // pass in the sun's specular emission

        gl.uniform3fv(moonAmbientULoc, moonAmbient); // pass in the moon's ambient emission
        gl.uniform3fv(moonDiffuseULoc, moonDiffuse); // pass in the moon's diffuse emission
        gl.uniform3fv(moonSpecularULoc, moonSpecular); // pass in the moon's specular emission

        gl.uniform3fv(ambientULoc, ambient); // pass in the ambient reflectivity
        gl.uniform3fv(this.diffuseULoc, diffuse); // pass in the diffuse reflectivity
        gl.uniform3fv(specularULoc, specular); // pass in the specular reflectivity
        gl.uniform1f(shininessULoc, shininess); // pass in the specular exponent
    }
}

// Load shaders as compile time constants
const vertexShader = require('./shader.vs').toString();
const fragmentShader = require('./shader.fs').toString();

export function createShader(gl: WebGLRenderingContext): Shader {
    return new Shader(gl, fragmentShader, vertexShader);
}